request/multiple_kernels.rst

```diff
@@ -0,0 +1,183 @@
+
+In most Symfony applications, incoming requests are processed by the
+``web/app.php`` front controller, which instantiates the ``app/AppKernel.php``
+class to create the application kernel that loads the bundles and handles the
+request to generate the response.
+
+This single kernel approach is a convenient default provided by the Symfony
+Standard edition, but Symfony applications can define any number of kernels.
+Whereas :doc:`environments </configuration/environments>` execute the same
+application with different configurations, kernels can execute different parts
+of the same application.
+
+These are some of the common use cases for creating multiple kernels:
+
+* An application that defines an API could define two kernels for performance
+  reasons. The first kernel would serve the regular application and the second
+  one would only respond to the API requests, loading less bundles and enabling
+  less features;
+* A highly sensitive application could define two kernels. The first one would
+  only load the routes that match the parts of the application exposed publicly.
+  The second kernel would load the rest of the application and its access would
+  be protected by the web server;
+* A micro-services oriented application could define several kernels to
+  enable/disable services selectively turning a traditional monolith application
+  into several micro-applications.
+
+Adding a new Kernel to the Application
+--------------------------------------
+
+Creating a new kernel in a Symfony application is a three-step process:
+
+1. Create a new front controller to load the new kernel;
+2. Create the new kernel class;
+3. Define the configuration loaded by the new kernel.
+
+The following example shows how to create a new kernel for the API of a given
+Symfony application.
+
+Step 1) Create a new Front Controller
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Instead of creating the new front controller from scratch, it's recommended to
+duplicate the existing ones. For example, create ``web/api_dev.php`` from
+``web/app_dev.php`` and ``web/api.php`` from ``web/app.php``.
+
+Then, update the code of the new front controllers to instantiate the new kernel
+class instead of the usual ``AppKernel`` class::
+
+    // web/api.php
+    // ...
+    $kernel = new ApiKernel('prod', false);
+    // ...
+
+    // web/api_dev.php
+    // ...
+    $kernel = new ApiKernel('dev', true);
+    // ...
+
+Step 2) Create the new Kernel Class
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now you need to define the ``ApiKernel`` class used by the new front controller.
+The recommendation again is to duplicate the existing ``app/AppKernel.php`` file
+and make the needed changes.
+
+In this example, the changes of the new ``ApiKernel`` would be to load less
+bundles than ``AppKernel`` and to change the location of the cache, logs and
+config files to not mess with the regular application::
+
+    // app/ApiKernel.php
+    <?php
```

Comment:
We usually don't have the `<?php`. If you *do* want it for some specific reason, the filename should probably be moved inside of it, so that the comment is in PHP. But probably, we should just remove `<?php`